alacritty_terminal/src/term/mod.rs

```diff
@@ -357,11 +357,7 @@ impl<T> Term<T> {
                 res += self
                     .line_to_string(line, start.column..end.column, start.column.0 != 0)
                     .trim_end();
-
-                // If the last column is included, newline is appended automatically.
-                if end.column != self.columns() - 1 {
-                    res += "\n";
-                }
+                res += "\n";
```

Comment:
We should probably revert to the old code and just trim whitespace?

Reply:
Oh, do you mean that we should trim whitespaces in `line_to_string` instead of `selection_to_string`? If so, is it for consistence among selections?

Reply:
No, I mean that with the change in 8e58409, we over-trimmed. Now we adjust that over-trimming by adding stuff again. That doesn't seem right to me. We should probably just not trim off stuff we don't want gone in the first place.

Reply:
Is non `over-trimming` like this?

```rust
// Trim just one newline and any number of other whitespace(except newline) on the end of line.
fn trim_end_weak(s: &str) -> String {
    let mut s = s.to_string();

    if s.ends_with('\n') {
        s.pop();
    }

    s.trim_end_matches(|c: char| c != '\n' && c.is_whitespace()).to_string()
}

#[test]
fn test_trim_end_weak() {
    assert_eq!(trim_end_weak("  hoge"), "  hoge".to_string());
    assert_eq!(trim_end_weak("  hoge\n"), "  hoge".to_string());
    assert_eq!(trim_end_weak("  hoge  "), "  hoge".to_string());
    assert_eq!(trim_end_weak("  hoge  \n"), "  hoge".to_string());
    assert_eq!(trim_end_weak("  hoge  \n  "), "  hoge  \n".to_string());
    assert_eq!(trim_end_weak("  hoge  \n\n"), "  hoge  \n".to_string());
    assert_eq!(trim_end_weak("  hoge  \n  \n"), "  hoge  \n".to_string());
    assert_eq!(trim_end_weak("  hoge  \n  \n  "), "  hoge  \n  \n".to_string());
}
```

Reply:
Honestly just `trem_end_matches(' ')` should work. We don't actually care about trimming anything other than empty cell content. If a user outputs text which contains whitespace that isn't actually empty terminal content, we should probably copy it anyway.

Reply:
Your comments probably make sense, but it is difficult to do it on the current implementation. Now no way to distinguish whether or not a cell is empty or just whitespace, right? Alacritty treats a space cell as a virtual empty cell on `line_to_string`. I think this makes sense on the current implementation.

BTW, it is awesome if we can distinguish an empty cell and a whitespace cell. It can improve the ideal behavior for the `FirstOccupied` action for vi-mode for me.

Reply:
> Now no way to distinguish whether or not a cell is empty or just whitespace, right?

Correct.

> BTW, it is awesome if we can distinguish an empty cell and a whitespace cell. It can improve the ideal behavior for the FirstOccupied action for vi-mode for me.

It isn't, especially because cursor motions allow writing in a different cell without actually ever inserting any whitespace. The only thing terminal emulators can realistically do is store what is an explicit and what is an implicit whitespace, but ultimately it's still all whitespace.